File: src/api/tauri-bridge/ai.ts
import { Channel, invoke } from "@tauri-apps/api/core"

import { call } from "./helper"

// region:    --- Models

export interface ModelInfo {
  name: string
  adapter_kind: string
}

export interface ProviderModels {
  provider: string
  models: ModelInfo[]
}

export async function listAllAiModels(): Promise<ProviderModels[]> {
  return await invoke<ProviderModels[]>("list_all_ai_models")
}

// endregion: --- Models

// region:    --- Chat Session

export interface ChatSession {
  id: string
  title: string
  model: string
  provider: string
  wid: number
  created_at: string | null
  updated_at: string | null
}

export interface ChatMessage {
  id: number
  sid: string
  role: string
  content: string
  created_at: string | null
}

export interface ChatSessionDetail {
  session: ChatSession
  messages: ChatMessage[]
}

export interface CreateChatSessionParams {
  wid: number
  provider: string
  model: string
  title?: string
}

export function createChatSession(params: CreateChatSessionParams): Promise<ChatSession> {
  return call<ChatSession>("create_chat_session", { req: params })
}

export function listChatSessions(wid: number): Promise<ChatSession[]> {
  return call<ChatSession[]>("list_chat_sessions", { wid })
}

export function getChatSessionData(sid: string): Promise<ChatSessionDetail> {
  return call<ChatSessionDetail>("get_chat_session_data", { sid })
}

export function deleteChatSession(sid: string): Promise<void> {
  return call<void>("delete_chat_session", { sid })
}

// endregion: --- Chat Session

// region:    --- Chat Stream

export type StreamEvent =
  | { type: "chunk"; content: string }
  | { type: "reasoning"; content: string }
  | { type: "done"; usage?: UsageInfo }
  | { type: "error"; message: string }

export interface UsageInfo {
  prompt_tokens?: number
  completion_tokens?: number
  total_tokens?: number
}

export interface ChatStreamParams {
  /** null/undefined 表示由后端 auto 选择 provider */
  provider: string | null
  /** null/undefined 表示由后端 auto 选择 model */
  model: string | null
  message: string
  /** 会话 id，必传 */
  sid: string
  /** Workspace the session belongs to; falls back to current workspace when omitted */
  wid?: number
}

export interface ChatStreamCallbacks {
  onChunk: (text: string) => void
  onReasoning?: (text: string) => void
  onDone?: (usage?: UsageInfo) => void
  onError?: (err: unknown) => void
}

export function chatStream(
  params: ChatStreamParams,
  callbacks: ChatStreamCallbacks
): Promise<void> {
  const { onChunk, onReasoning, onDone, onError } = callbacks

  const channel = new Channel<StreamEvent>()
  channel.onmessage = (evt: StreamEvent) => {
    switch (evt.type) {
      case "chunk":
        onChunk(evt.content)
        break
      case "reasoning":
        onReasoning?.(evt.content)
        break
      case "done":
        onDone?.(evt.usage)
        break
      case "error":
        onError?.(evt.message)
        break
    }
  }

  return invoke<void>("chat_stream", { req: params, channel }).catch((err: unknown) => {
    onError?.(err)
    return Promise.reject(err)
  })
}

export function abortStream(sid: string): Promise<void> {
  return invoke<void>("abort_stream", { sid })
}

// endregion: --- Chat Stream

File: src-tauri/src/interfaces/dto/ai.rs
use serde::{Deserialize, Serialize};

use crate::domain::ai::entity::{AiChatMessage, AiChatSession, AiModel, AiSessionData};
use crate::domain::ai::repo::ProviderModels;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelInfoDto {
    pub name: String,
    pub adapter_kind: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ProviderModelsDto {
    pub provider: String,
    pub models: Vec<ModelInfoDto>,
}

impl From<AiModel> for ModelInfoDto {
    fn from(m: AiModel) -> Self {
        Self {
            name: m.name,
            adapter_kind: m.adapter_kind,
        }
    }
}

impl From<ProviderModels> for ProviderModelsDto {
    fn from(p: ProviderModels) -> Self {
        Self {
            provider: p.provider.as_str().to_string(),
            models: p.models.into_iter().map(ModelInfoDto::from).collect(),
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct ChatStreamReq {
    pub provider: Option<String>,
    pub model: Option<String>,
    pub message: String,
    pub sid: String,
    pub wid: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StreamEvent {
    Chunk { content: String },
    Reasoning { content: String },
    Done { usage: Option<UsageDto> },
    Error { message: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageDto {
    pub prompt_tokens: Option<i32>,
    pub completion_tokens: Option<i32>,
    pub total_tokens: Option<i32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatSessionDto {
    pub id: String,
    pub title: String,
    pub model: String,
    pub provider: String,
    pub wid: i64,
    pub created_at: Option<chrono::NaiveDateTime>,
    pub updated_at: Option<chrono::NaiveDateTime>,
}

impl From<AiChatSession> for ChatSessionDto {
    fn from(s: AiChatSession) -> Self {
        Self {
            id: s.id,
            title: s.title,
            model: s.model,
            provider: s.provider,
            wid: s.wid,
            created_at: s.created_at,
            updated_at: s.updated_at,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatMessageDto {
    pub id: i64,
    pub sid: String,
    pub role: String,
    pub content: String,
    pub created_at: Option<chrono::NaiveDateTime>,
}

impl From<AiChatMessage> for ChatMessageDto {
    fn from(m: AiChatMessage) -> Self {
        Self {
            id: m.id,
            sid: m.sid,
            role: m.role,
            content: m.content,
            created_at: m.created_at,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatSessionDetailDto {
    pub session: ChatSessionDto,
    pub messages: Vec<ChatMessageDto>,
}

impl From<AiSessionData> for ChatSessionDetailDto {
    fn from(d: AiSessionData) -> Self {
        Self {
            session: d.session.into(),
            messages: d.messages.into_iter().map(ChatMessageDto::from).collect(),
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateChatSessionReq {
    pub wid: i64,
    pub provider: String,
    pub model: String,
    pub title: Option<String>,
}
